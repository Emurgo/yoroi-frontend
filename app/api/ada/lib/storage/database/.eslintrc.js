module.exports = {
  rules: {
    // disable member variable rule since it doesn't work well with the pattern we use
    "no-restricted-syntax": [0],
  },
}
