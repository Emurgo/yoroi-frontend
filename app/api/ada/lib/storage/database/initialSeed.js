// @flow

import crypto from 'crypto';

export function getInitialSeeds(): {|
  AddressSeed: number,
  TransactionSeed: number,
  BlockSeed: number,
  |} {
  return {
    AddressSeed: crypto.randomBytes(4).readUInt32BE(0),
    TransactionSeed: crypto.randomBytes(4).readUInt32BE(0),
    BlockSeed: crypto.randomBytes(4).readUInt32BE(0),
  };
}
