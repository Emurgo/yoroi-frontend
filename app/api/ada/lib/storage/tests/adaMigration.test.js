// @flow

import '../../test-config';
import { schema, } from 'lovefield';
import type { lf$Database } from 'lovefield';
import { setLocalItem } from '../../../../localStorage/primitives';
import oldStorageMemory from '../../../../../../features/yoroi_snapshots/historical-versions/1_9_0/software/localStorage.json';
import oldStorageTrezor from '../../../../../../features/yoroi_snapshots/historical-versions/1_9_0/trezor/localStorage.json';
import oldStorageLedger from '../../../../../../features/yoroi_snapshots/historical-versions/1_9_0/ledger/localStorage.json';
import { RustModule } from '../../cardanoCrypto/rustLoader';
import { dumpByVersion, loadLovefieldDB } from '../database/index';
import { storagev2Migation } from '../adaMigration';
import { mockDate, filterDbSnapshot } from '../bridge/tests/common';

beforeAll(async () => {
  await RustModule.load();
});

beforeEach(() => {
  mockDate();
});

async function baseTest(
  db: lf$Database,
): Promise<void> {
  // need to fake having data in the legacy DB format for migration to trigger
  dumpByVersion.test = [];

  await storagev2Migation(db);

  const keysForTest = [
    'ConceptualWallet',
    'Key',
    'Bip44Wrapper',
    'RootDerivation',
    'PurposeDerivation',
    'CoinTypeDerivation',
    'Bip44Account',
    'Bip44Chain',
    'HwWalletMeta',
    'KeyDerivation',
  ];

  const dump = (await db.export()).tables;
  filterDbSnapshot(dump, keysForTest);
}

test('Migrate memory storage v1 to storage v2', async (done) => {
  const db = await loadLovefieldDB(schema.DataStoreType.MEMORY);
  for (const key of Object.keys(oldStorageMemory)) {
    await setLocalItem(key, oldStorageMemory[key]);
  }

  await baseTest(db);
  done();
});

test('Migrate trezor storage v1 to storage v2', async (done) => {
  const db = await loadLovefieldDB(schema.DataStoreType.MEMORY);
  for (const key of Object.keys(oldStorageTrezor)) {
    await setLocalItem(key, oldStorageTrezor[key]);
  }

  await baseTest(db);
  done();
});

test('Migrate ledger storage v1 to storage v2', async (done) => {
  const db = await loadLovefieldDB(schema.DataStoreType.MEMORY);
  for (const key of Object.keys(oldStorageLedger)) {
    await setLocalItem(key, oldStorageLedger[key]);
  }

  await baseTest(db);
  done();
});
