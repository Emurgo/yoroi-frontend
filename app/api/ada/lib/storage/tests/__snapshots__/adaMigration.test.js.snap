// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrate ledger storage v1 to storage v2 1`] = `
Array [
  Object {
    "Bip44Account": Array [
      Object {
        "Bip44AccountId": 1,
        "KeyDerivationId": 4,
      },
    ],
  },
  Object {
    "Bip44Chain": Array [
      Object {
        "Bip44ChainId": 1,
        "DisplayCutoff": 77,
        "KeyDerivationId": 5,
      },
      Object {
        "Bip44ChainId": 2,
        "DisplayCutoff": null,
        "KeyDerivationId": 26,
      },
    ],
  },
  Object {
    "Bip44Wrapper": Array [
      Object {
        "Bip44WrapperId": 1,
        "ConceptualWalletId": 1,
        "PrivateDeriverKeyDerivationId": null,
        "PrivateDeriverLevel": null,
        "PublicDeriverLevel": 3,
        "RootKeyDerivationId": 1,
        "SignerLevel": null,
      },
    ],
  },
  Object {
    "CoinTypeDerivation": Array [
      Object {
        "CoinTypeDerivationId": 1,
        "KeyDerivationId": 3,
      },
    ],
  },
  Object {
    "ConceptualWallet": Array [
      Object {
        "CoinType": 2147485463,
        "ConceptualWalletId": 1,
        "Name": "Yoroi-Ledger",
      },
    ],
  },
  Object {
    "HwWalletMeta": Array [
      Object {
        "ConceptualWalletId": 1,
        "DeviceId": "",
        "HwWalletMetaId": 1,
        "Label": "",
        "Language": "",
        "MajorVersion": 1,
        "MinorVersion": 0,
        "Model": "NanoS",
        "PatchVersion": 0,
        "Vendor": "ledger.com",
      },
    ],
  },
  Object {
    "Key": Array [
      Object {
        "Hash": "76bca341ea86128125b49999ea1799fcd0d005a64b9150c33ed4648a3c11b20c26ea85882a1616bbf4ffae3fae6debe77169a3b299ec226409e2def1bffd35a9",
        "IsEncrypted": false,
        "KeyId": 1,
        "PasswordLastUpdate": null,
      },
    ],
  },
  Object {
    "KeyDerivation": Array [
      Object {
        "Index": null,
        "KeyDerivationId": 1,
        "Parent": null,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147483692,
        "KeyDerivationId": 2,
        "Parent": 1,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147485463,
        "KeyDerivationId": 3,
        "Parent": 2,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147483648,
        "KeyDerivationId": 4,
        "Parent": 3,
        "PrivateKeyId": null,
        "PublicKeyId": 1,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 5,
        "Parent": 4,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 6,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 7,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2,
        "KeyDerivationId": 8,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 3,
        "KeyDerivationId": 9,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 4,
        "KeyDerivationId": 10,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 5,
        "KeyDerivationId": 11,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 6,
        "KeyDerivationId": 12,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 7,
        "KeyDerivationId": 13,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 8,
        "KeyDerivationId": 14,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 9,
        "KeyDerivationId": 15,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 10,
        "KeyDerivationId": 16,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 11,
        "KeyDerivationId": 17,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 12,
        "KeyDerivationId": 18,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 13,
        "KeyDerivationId": 19,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 14,
        "KeyDerivationId": 20,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 15,
        "KeyDerivationId": 21,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 16,
        "KeyDerivationId": 22,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 17,
        "KeyDerivationId": 23,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 18,
        "KeyDerivationId": 24,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 19,
        "KeyDerivationId": 25,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 26,
        "Parent": 4,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 27,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 28,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2,
        "KeyDerivationId": 29,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 3,
        "KeyDerivationId": 30,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 4,
        "KeyDerivationId": 31,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 5,
        "KeyDerivationId": 32,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 6,
        "KeyDerivationId": 33,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 7,
        "KeyDerivationId": 34,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 8,
        "KeyDerivationId": 35,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 9,
        "KeyDerivationId": 36,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 10,
        "KeyDerivationId": 37,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 11,
        "KeyDerivationId": 38,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 12,
        "KeyDerivationId": 39,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 13,
        "KeyDerivationId": 40,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 14,
        "KeyDerivationId": 41,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 15,
        "KeyDerivationId": 42,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 16,
        "KeyDerivationId": 43,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 17,
        "KeyDerivationId": 44,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 18,
        "KeyDerivationId": 45,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 19,
        "KeyDerivationId": 46,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
    ],
  },
  Object {
    "PurposeDerivation": Array [
      Object {
        "KeyDerivationId": 2,
        "PurposeDerivationId": 1,
      },
    ],
  },
  Object {
    "RootDerivation": Array [
      Object {
        "KeyDerivationId": 1,
        "RootDerivationId": 1,
      },
    ],
  },
]
`;

exports[`Migrate memory storage v1 to storage v2 1`] = `
Array [
  Object {
    "Bip44Account": Array [
      Object {
        "Bip44AccountId": 1,
        "KeyDerivationId": 4,
      },
    ],
  },
  Object {
    "Bip44Chain": Array [
      Object {
        "Bip44ChainId": 1,
        "DisplayCutoff": 142,
        "KeyDerivationId": 5,
      },
      Object {
        "Bip44ChainId": 2,
        "DisplayCutoff": null,
        "KeyDerivationId": 26,
      },
    ],
  },
  Object {
    "Bip44Wrapper": Array [
      Object {
        "Bip44WrapperId": 1,
        "ConceptualWalletId": 1,
        "PrivateDeriverKeyDerivationId": 1,
        "PrivateDeriverLevel": 0,
        "PublicDeriverLevel": 3,
        "RootKeyDerivationId": 1,
        "SignerLevel": 0,
      },
    ],
  },
  Object {
    "CoinTypeDerivation": Array [
      Object {
        "CoinTypeDerivationId": 1,
        "KeyDerivationId": 3,
      },
    ],
  },
  Object {
    "ConceptualWallet": Array [
      Object {
        "CoinType": 2147485463,
        "ConceptualWalletId": 1,
        "Name": "Wallet",
      },
    ],
  },
  Object {
    "HwWalletMeta": Array [],
  },
  Object {
    "Key": Array [
      Object {
        "Hash": "a4f667dd7dd4e242fdc082540f24996a54227f6bbf0e77f735fc192fca4dbe8fcad69edddcd0ab1c5d1db62717c6b733fba5f19bcf4cda7a17770408342c4eb25373ab7e47678ab20ad428301a39324a0f789f10901ce41a128b466568f863b9b7baa989aaf09ec746615ea2400430dadec18ada015841107bd2133dc3b2af107ecea49c1b345bdb51ebda9fed6ebe6546efa20524d2dcc66c8a5605",
        "IsEncrypted": true,
        "KeyId": 1,
        "PasswordLastUpdate": 2019-07-31T10:40:16.000Z,
      },
      Object {
        "Hash": "b2938f29eb965952b03baaf010b0f5fe4a2c116a1e255274c9ace24deb72450ce1c045821c42ed11b33db3124b1551793e76b8ad187efb92b28969b92bb95964",
        "IsEncrypted": false,
        "KeyId": 2,
        "PasswordLastUpdate": null,
      },
    ],
  },
  Object {
    "KeyDerivation": Array [
      Object {
        "Index": null,
        "KeyDerivationId": 1,
        "Parent": null,
        "PrivateKeyId": 1,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147483692,
        "KeyDerivationId": 2,
        "Parent": 1,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147485463,
        "KeyDerivationId": 3,
        "Parent": 2,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147483648,
        "KeyDerivationId": 4,
        "Parent": 3,
        "PrivateKeyId": null,
        "PublicKeyId": 2,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 5,
        "Parent": 4,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 6,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 7,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2,
        "KeyDerivationId": 8,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 3,
        "KeyDerivationId": 9,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 4,
        "KeyDerivationId": 10,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 5,
        "KeyDerivationId": 11,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 6,
        "KeyDerivationId": 12,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 7,
        "KeyDerivationId": 13,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 8,
        "KeyDerivationId": 14,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 9,
        "KeyDerivationId": 15,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 10,
        "KeyDerivationId": 16,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 11,
        "KeyDerivationId": 17,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 12,
        "KeyDerivationId": 18,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 13,
        "KeyDerivationId": 19,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 14,
        "KeyDerivationId": 20,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 15,
        "KeyDerivationId": 21,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 16,
        "KeyDerivationId": 22,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 17,
        "KeyDerivationId": 23,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 18,
        "KeyDerivationId": 24,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 19,
        "KeyDerivationId": 25,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 26,
        "Parent": 4,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 27,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 28,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2,
        "KeyDerivationId": 29,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 3,
        "KeyDerivationId": 30,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 4,
        "KeyDerivationId": 31,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 5,
        "KeyDerivationId": 32,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 6,
        "KeyDerivationId": 33,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 7,
        "KeyDerivationId": 34,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 8,
        "KeyDerivationId": 35,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 9,
        "KeyDerivationId": 36,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 10,
        "KeyDerivationId": 37,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 11,
        "KeyDerivationId": 38,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 12,
        "KeyDerivationId": 39,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 13,
        "KeyDerivationId": 40,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 14,
        "KeyDerivationId": 41,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 15,
        "KeyDerivationId": 42,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 16,
        "KeyDerivationId": 43,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 17,
        "KeyDerivationId": 44,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 18,
        "KeyDerivationId": 45,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 19,
        "KeyDerivationId": 46,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
    ],
  },
  Object {
    "PurposeDerivation": Array [
      Object {
        "KeyDerivationId": 2,
        "PurposeDerivationId": 1,
      },
    ],
  },
  Object {
    "RootDerivation": Array [
      Object {
        "KeyDerivationId": 1,
        "RootDerivationId": 1,
      },
    ],
  },
]
`;

exports[`Migrate trezor storage v1 to storage v2 1`] = `
Array [
  Object {
    "Bip44Account": Array [
      Object {
        "Bip44AccountId": 1,
        "KeyDerivationId": 4,
      },
    ],
  },
  Object {
    "Bip44Chain": Array [
      Object {
        "Bip44ChainId": 1,
        "DisplayCutoff": 61,
        "KeyDerivationId": 5,
      },
      Object {
        "Bip44ChainId": 2,
        "DisplayCutoff": null,
        "KeyDerivationId": 26,
      },
    ],
  },
  Object {
    "Bip44Wrapper": Array [
      Object {
        "Bip44WrapperId": 1,
        "ConceptualWalletId": 1,
        "PrivateDeriverKeyDerivationId": null,
        "PrivateDeriverLevel": null,
        "PublicDeriverLevel": 3,
        "RootKeyDerivationId": 1,
        "SignerLevel": null,
      },
    ],
  },
  Object {
    "CoinTypeDerivation": Array [
      Object {
        "CoinTypeDerivationId": 1,
        "KeyDerivationId": 3,
      },
    ],
  },
  Object {
    "ConceptualWallet": Array [
      Object {
        "CoinType": 2147485463,
        "ConceptualWalletId": 1,
        "Name": "Trezorro",
      },
    ],
  },
  Object {
    "HwWalletMeta": Array [
      Object {
        "ConceptualWalletId": 1,
        "DeviceId": "Trezorro",
        "HwWalletMetaId": 1,
        "Label": "C875BA9D0C571FF4B8718FAA",
        "Language": "english",
        "MajorVersion": 2,
        "MinorVersion": 1,
        "Model": "T",
        "PatchVersion": 1,
        "Vendor": "trezor.io",
      },
    ],
  },
  Object {
    "Key": Array [
      Object {
        "Hash": "bc772da2ba5a37bad47f51d6a4476b236aa6734b5cd79c81706d5018d3203f6c05d510314991bfd73c2a3e9da72d2e1205b54cf6342f588a4fbbaaa98e58211b",
        "IsEncrypted": false,
        "KeyId": 1,
        "PasswordLastUpdate": null,
      },
    ],
  },
  Object {
    "KeyDerivation": Array [
      Object {
        "Index": null,
        "KeyDerivationId": 1,
        "Parent": null,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147483692,
        "KeyDerivationId": 2,
        "Parent": 1,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147485463,
        "KeyDerivationId": 3,
        "Parent": 2,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2147483648,
        "KeyDerivationId": 4,
        "Parent": 3,
        "PrivateKeyId": null,
        "PublicKeyId": 1,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 5,
        "Parent": 4,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 6,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 7,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2,
        "KeyDerivationId": 8,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 3,
        "KeyDerivationId": 9,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 4,
        "KeyDerivationId": 10,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 5,
        "KeyDerivationId": 11,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 6,
        "KeyDerivationId": 12,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 7,
        "KeyDerivationId": 13,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 8,
        "KeyDerivationId": 14,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 9,
        "KeyDerivationId": 15,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 10,
        "KeyDerivationId": 16,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 11,
        "KeyDerivationId": 17,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 12,
        "KeyDerivationId": 18,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 13,
        "KeyDerivationId": 19,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 14,
        "KeyDerivationId": 20,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 15,
        "KeyDerivationId": 21,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 16,
        "KeyDerivationId": 22,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 17,
        "KeyDerivationId": 23,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 18,
        "KeyDerivationId": 24,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 19,
        "KeyDerivationId": 25,
        "Parent": 5,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 26,
        "Parent": 4,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 0,
        "KeyDerivationId": 27,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 1,
        "KeyDerivationId": 28,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 2,
        "KeyDerivationId": 29,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 3,
        "KeyDerivationId": 30,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 4,
        "KeyDerivationId": 31,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 5,
        "KeyDerivationId": 32,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 6,
        "KeyDerivationId": 33,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 7,
        "KeyDerivationId": 34,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 8,
        "KeyDerivationId": 35,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 9,
        "KeyDerivationId": 36,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 10,
        "KeyDerivationId": 37,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 11,
        "KeyDerivationId": 38,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 12,
        "KeyDerivationId": 39,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 13,
        "KeyDerivationId": 40,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 14,
        "KeyDerivationId": 41,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 15,
        "KeyDerivationId": 42,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 16,
        "KeyDerivationId": 43,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 17,
        "KeyDerivationId": 44,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 18,
        "KeyDerivationId": 45,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
      Object {
        "Index": 19,
        "KeyDerivationId": 46,
        "Parent": 26,
        "PrivateKeyId": null,
        "PublicKeyId": null,
      },
    ],
  },
  Object {
    "PurposeDerivation": Array [
      Object {
        "KeyDerivationId": 2,
        "PurposeDerivationId": 1,
      },
    ],
  },
  Object {
    "RootDerivation": Array [
      Object {
        "KeyDerivationId": 1,
        "RootDerivationId": 1,
      },
    ],
  },
]
`;
